torch/utils/data/dataloader.py

```diff
@@ -103,13 +103,15 @@ def default_collate(batch):
         return torch.DoubleTensor(batch)
     elif isinstance(batch[0], string_classes):
         return batch
-    elif isinstance(batch[0], collections.Iterable):
-        # if each batch element is not a tensor, then it should be a tuple
-        # of tensors; in that case we collate each element in the tuple
+    elif isinstance(batch[0], collections.Mapping):
+        return {key: default_collate([d[key] for d in batch]) for key in batch[0]}
+    elif isinstance(batch[0], collections.Sequence):
```

Comment:
Dataloder should raise an error with sets (set are unordered) as batching sets is ambiguous.

Example code:

```
import torch
from torch.utils.data import DataLoader, Dataset


class SetDataset(Dataset):
    def __len__(self):
        return 12

    def __getitem__(self, idx):
        return {torch.Tensor(3, 1, 3).fill_(idx),
                torch.Tensor(4, 3).fill_(idx),
                torch.Tensor(5, 3, 3).fill_(idx),
                torch.Tensor(5, 3).fill_(idx),
                torch.Tensor(3).fill_(idx),
                }


for sample in DataLoader(SetDataset(), 4, shuffle=False):
    pass
```

Error from my version:
```Traceback (most recent call last):
  File "test_lol.py", line 18, in <module>
    for sample in DataLoader(SetDataset(), 4, shuffle=False):
  File "/Users/Sasank/anaconda3/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 192, in __next__
    batch = self.collate_fn([self.dataset[i] for i in indices])
  File "/Users/Sasank/anaconda3/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 115, in default_collate
    .format(type(batch[0]))))
TypeError: batch must contain tensors, numbers, dicts or lists; found <class 'set'>
```

from existing version:
```
Traceback (most recent call last):
  File "test_lol.py", line 18, in <module>
    for sample in DataLoader(SetDataset(), 4, shuffle=False):
  File "/home/users/sasank/anaconda3/lib/python3.5/site-packages/torch/utils/data/dataloader.py", line 152, in __next__
    batch = self.collate_fn([self.dataset[i] for i in indices])
  File "/home/users/sasank/anaconda3/lib/python3.5/site-packages/torch/utils/data/dataloader.py", line 79, in default_collate
    return [default_collate(samples) for samples in transposed]
  File "/home/users/sasank/anaconda3/lib/python3.5/site-packages/torch/utils/data/dataloader.py", line 79, in <listcomp>
    return [default_collate(samples) for samples in transposed]
  File "/home/users/sasank/anaconda3/lib/python3.5/site-packages/torch/utils/data/dataloader.py", line 66, in default_collate
    return torch.stack(batch, 0)
  File "/home/users/sasank/anaconda3/lib/python3.5/site-packages/torch/functional.py", line 56, in stack
    return torch.cat(list(t.unsqueeze(dim) for t in sequence), dim)
RuntimeError: inconsistent tensor sizes at /py/conda-bld/pytorch_1490983232023/work/torch/lib/TH/generic/THTensorMath.c:2548
```